File: Meetrix/app/src/main/java/com/starrepublic/meetrix/widget/SnappingHorizontalScrollView.java
package com.starrepublic.meetrix.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.HorizontalScrollView;
import android.widget.OverScroller;
import android.widget.Scroller;

import java.lang.reflect.Field;

/**
 * Created by richard on 2016-10-30.
 */

public class SnappingHorizontalScrollView extends HorizontalScrollView {
    private static final String TAG = SnappingHorizontalScrollView.class.getSimpleName();
    private boolean currentlyTouching;
    private boolean currentlyScrolling;
    private int lastDiff = -1;

    private int snapTo = 0;


    public SnappingHorizontalScrollView(Context context) {
        super(context);
        init();
    }



    public SnappingHorizontalScrollView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SnappingHorizontalScrollView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public SnappingHorizontalScrollView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    private void init() {
    }




    @Override
    protected void onScrollChanged(int x, int y, int oldX, int oldY) {

        int diff = Math.abs(x - oldX);
        if (diff > 1) {
            currentlyScrolling = true;
            lastDiff = -1;
        } else {
            currentlyScrolling = false;
            if (!currentlyTouching && lastDiff != diff) {
                lastDiff = diff;
                scrollStopped();
            }
        }
        super.onScrollChanged(x, y, oldX, oldY);
    }



    private void scrollStopped() {
        //Log.d(TAG, "SCROLL STOPPED!!!");

        //smoothScrollTo(snapTo*Math.round(getScrollX()/(float)snapTo),0);

    }





    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                currentlyTouching = true;
        }
        return super.onInterceptTouchEvent(event);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                currentlyTouching = false;
                if (!currentlyScrolling) {
                    scrollStopped();
                    return super.onTouchEvent(event);
                }
        }
        return super.onTouchEvent(event);
    }

    public int getSnapTo() {
        return snapTo;
    }

    public void setSnapTo(int snapTo) {
        this.snapTo = snapTo;
    }
}

File: Meetrix/app/src/main/java/com/starrepublic/meetrix/widget/SnappingHorizontalScrollView.kt
package com.starrepublic.meetrix.widget;

import android.content.Context
import android.hardware.SensorManager
import android.support.v4.view.GestureDetectorCompat
import android.support.v4.view.MotionEventCompat
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ViewConfiguration
import android.widget.HorizontalScrollView

/**
 * Created by richard on 2016-10-30.
 */
class SnappingHorizontalScrollView : HorizontalScrollView {

    constructor(context: Context) : this(context, null)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr, 0)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        detector = GestureDetectorCompat(context, GestureListener())
    }

    private val detector: GestureDetectorCompat
    private var currentlyTouching: Boolean = false;
    private var currentlyScrolling: Boolean = false
    private var lastDiff: Int = -1;
    var snappingEnabled: Boolean = false;
    var snapTo: Int = 0;
    var scrollable: Boolean = true
    fun scrollStopped() {
        if (snappingEnabled) {
            handler?.post {
                smoothScrollTo(snapTo * Math.round(scrollX / snapTo.toFloat()), 0)
            }
        }
    }

    override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean {
        if (!scrollable) return false
        return super.onInterceptTouchEvent(ev)
    }

    private var flinging: Boolean = false
    override fun onTouchEvent(event: MotionEvent?): Boolean {
        if (!scrollable) return false
        this.detector.onTouchEvent(event)
        val action = MotionEventCompat.getActionMasked(event)

        when (action) {
            MotionEvent.ACTION_DOWN -> {
                flinging = false
                handler.removeCallbacks(resetRunnable)
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                if (!flinging) {
                    scrollStopped()
                }
            }
        }

        return super.onTouchEvent(event)
    }

    private val resetRunnable: Runnable = Runnable { scrollStopped() }

    internal inner class GestureListener : GestureDetector.SimpleOnGestureListener() {
        private val INFLEXION = 0.35f
        private val DECELERATION_RATE = Math.log(0.78) / Math.log(0.9)
        private val ppi: Float
        private val physicalCoeff: Float
        private val flingFriction = ViewConfiguration.getScrollFriction()

        init {
            ppi = context.resources.displayMetrics.density * 160.0f
            physicalCoeff = computeDeceleration(0.84f)
        }

        private fun computeDeceleration(friction: Float): Float {
            return SensorManager.GRAVITY_EARTH * 39.37f * ppi * friction
        }

        private fun getSplineDeceleration(velocity: Float): Double {
            return Math.log((INFLEXION * Math.abs(velocity) / (flingFriction * physicalCoeff)).toDouble())
        }

        override fun onFling(event1: MotionEvent?, event2: MotionEvent?, velocityX: Float, velocityY: Float): Boolean {
            flinging = true
            val l = getSplineDeceleration(velocityX);
            val decelMinusOne = DECELERATION_RATE - 1.0;
            val duration = (1000.0 * Math.exp(l / decelMinusOne)).toLong()
            handler.postDelayed(resetRunnable, duration)
            return false
        }

        override fun onScroll(event1: MotionEvent?, event2: MotionEvent?, distanceX: Float, distanceY: Float): Boolean {
            return false
        }
    }
}
